### DISCUSSION TITLE
HEX/Byte to ASCII in Micro Python

### DISCUSSION THREAD

**Discussion by pierregraciano** [micropython/micropython#13575] [Using MicroPython]
Hello! Im trying to convert a var to ASCII in uPy but didnt work. I read the var from my UART Port and this var has a byte value, and I need to convert this to ASCII. I tried to use .decode('ascii') but failed. 

Master code:
```
data_slave = uart.read()           
        if data_slave:
            print(f"Message Slave {i+1}: {data_slave}\n")
```

Slave code: 
```
data_sent = b"Hello (SLAVE1)"
uart.write(data_sent)
```

Console:
b'\x00\xc8\x00\xf0\x00\xfa\x00\xfe\x00\xff\x00\xfe\x00\x00\xff\x00\xfe\x00\xfe\x00\xff\x00\xfc\x00\xff\x00\x00\xf8\x00\xff\x00\xff\x00\xfc\x00\xfc\x00\xf8\x00\xfc\x00\xfc\x00\xff\x00\xff\x00\xff\x00\xfc\x00\xff\x00\xfc\x00\xfe\x00\xfe\x00\xfe*Y\x01\x00\xfc\x00\xfc\x00\xff\x00\xff\x00\xf8\x00\xfe\x00\xfc\x00\xff\x00\xff\x00\xfe\x01\xfe\x00\xff\x00\xf8\x00\xff\x00\xfe\x00\xfe\x00\xff\x00\xfe\x00\xf8\x00\xfc\x00\xfe\x00\xfc\x00\xf0\x00\xfe\x00\xfe\x00\xfe\x00\xf8\x00\xfe\x00\xfe\x00\xfe\x00\xfc\x00\xfe\x00\xfc\x00\xfe\x00\xfe\x00\xfe\x00\xfe\x00\xf8\x00\xfc\x00\xff\x00\xfc\x00\xfc\x00\xf8\x00\xfc\x00\xfc\x00\xfc\x00\xfc\x00\xfc\x00\xf8\x00\xf8\x00\xf8\x00\xfe\x00\xfe\x00\xfe\x00\xfc\x00\xfe\x00\x00\xff\x00\xfc\x00'

**Comment by robert-hh** [comment]
```
import binascii
binascii.hexlify(data_sent)
```
or as shorter output in base64 encoding:
```
import binascii
binascii.b2a_base64(data_sent)

```

**Reply by pierregraciano** [reply]
Now the console shows:

b'0094ff00fe00fe0000fe00ff00ff00ff00ff00ff00fe00fc00fa00fc00fc00ff0000fc00fe00fc0000f800fc00fc00fe00fc00fe00fe00f800fc001020c500f800e000f800fe00ff00f400fe0000fc00f800fe00f800fe00fc00f000fe00f800fc00fe00f800fe00fc00fc00fc00ff00fe00f400fc00f800fe00f000f800f800fe00f800fe00ff00f800f000fc00ff00f800f000fe00f800f400f800f800f000ff00f800fe00fc00ff00fe00f800fe00fc00fc'

And should show: b"Hello (SLAVE1)"

**Comment by robert-hh** [comment]
There is something else wrong with your connection, not the encoding. Baudrate? Missing GND wire?

**Reply by pierregraciano** [reply]
Baudrate Im using 115200 both of them and all ESPs conected on same GND. May I try to change the baudrate?

**Reply by robert-hh** [reply]
The data you show is just rubbish, picking up noise. 115200 Baud should be fine on a short wire (<5m). Double check the baud rates and the wiring. GND to GND, TXa to RXb, RXa to TXb.

**Reply by pierregraciano** [reply]
Now Im using buffers 3-state to accept or not the communication RX/TX - TX/RX of Slave and Master. Its possible the problem be the noise of buffer? Im using 20cm jumpers, so the distance for me is good

**Reply by robert-hh** [reply]
Use pull-up resistors on RX. With 3-state outputs the input is floating, when the buffer is switched off. For UART init, you can put everything in 1 statement `uart = UART(2, 115200, tx=17, rx=16)`.

**Reply by pierregraciano** [reply]
Thank you, Robert! Ill try again here!

**Reply by pierregraciano** [reply]
Robert, after put 10KOhm pull-up resistors my UART is not reading rubbish anymore, but I cant send and receive data using UART2.

**Reply by robert-hh** [reply]
That certainly unveils the fact that the set-up never worked with 2 slaves. Can you show us how you connected the boards together. @shariltumin comments below, that two TX pins are joined. That will not work. But you mentioned the 3-state buffers.is it valid to assume that the TX pins of the slave devices are fed through the 3-state buffers to the master? Then, you have to consider, that the master will only receive data as long as a certain slave is enabled. You you might need a protocol, where the master sends a request and the slave answers only then.

b.t.w.. Since you use ESP32s, what stops you from using WiFi for the communication? Then you have no wiring issue, and you can have a large number of slaves. Collisions are handled by the WiFi stack.

**Reply by pierregraciano** [reply]
So, the buffer open/close the circuit between TXa / RXb and TXb / RXa. Im using handshake, so the master send "OK1" to slave, the slave read the data and send "ACK1" to master. To control the buffer, master send GPIO signal to ports of control, so when I need the communication, the master "open" the buffer and now the ESPs will exchange data. Now I can send OK to slave and the slave can send the ACK but my master is reading nothing. I'm using a board to connect them, but today I'm going to solder the cables and see if it works. 
About WiFi, my system requires that use of wires.

**Reply by robert-hh** [reply]
Do you have a logic analyzer to tell, whether data arrives at the RX pin. Even a simple $10 device will do the job. If you have a spare RPi Pico, you can as well use logic analyzer software with the Pico, like this one https://github.com/dotcypress/ula

**Reply by pierregraciano** [reply]
Thanks Robert! I just reorganize the wires and change my testboard and my system read the data for the first time. After this it read rubbish again but I guess is the first step.

**Comment by shariltumin** [comment]
I'm interested in this. Could you please share your circuit diagram of 1 master and 2 slaves sharing a UART?  

Assuming that this is indeed the case, both TX lines of the 2 slaves are connected to the RX of the master. Won't these two TXs interfere with each other and just produce noise to the master RX?

It is not clear to me how the control_buffers pins are used to control the flow of data from the slaves and the master. Maybe I'm missing something here.

**Comment by shariltumin** [comment]
I've been playing around with multi-host UART communication. For those interested you can find my results [here](https://github.com/shariltumin/Multi-Hosts-UART).
